File: 2018/d.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# 2018 ACM-ICPC国内予選
# Problem D: 全チームによるプレーオフ

def product(m, n):
    p = 1
    for i in range(m, n+1): p *= i
    return p
def choose(n, k):
    return product(n-k+1, n) // product(1, k)

def solve(n, match):
    '''param n チーム数
       param match 実施済試合結果
       return 全プレーオフになる組合せ数'''
    global table, nwon, nlost, nmatch
    # 実施済試合数
    nmatch = [0] * n
    for x, y in match:
        nmatch[x] += 1
        nmatch[y] += 1
    # 実施済試合数の多い順にチーム番号を付け直す
    teams = sorted(range(n), key=lambda x: -nmatch[x])
    trans = [0] * n
    for i in range(n): trans[teams[i]] = i
    nmatch = [nmatch[teams[i]] for i in range(n)]
    # 星取表
    table = [[None for a in range(n)] for t in range(n)]
    for x, y in match:
        u, v = trans[x], trans[y]
        table[u][v] = 1  # won
        table[v][u] = 0  # lost
    nwon, nlost = [[sum(1 for y in range(n) if table[x][y] == i)
                    for x in range(n)] for i in range(2)]
    if not all(valid(x) for x in range(n)):
        return 0  # 初期配置が条件を満たさない
    return search(pos=0)

def valid(x):
    '''チームxの勝数・負数が(n-1)//2以下か'''
    global table, nwon, nlost
    n = len(table)
    return nwon[x] <= (n-1)//2 and nlost[x] <= (n-1)//2

def search(pos):
    '''param pos table中の次に埋める位置
       return 全プレーオフになる組合せ数'''
    global table, nwon, nlost, nmatch
    n = len(table)
    x = pos // n
    y = pos %  n
    if pos >= n * n:
        return 1
    if x == y or table[x][y] != None:
        return search(pos + 1)

    # 1試合もしていないチームは1パタンだけ試して組合せ倍する
    if x == 0 and 0 < y and nmatch[y] == 0:
        i = 0 if nwon[x] < (n-1)//2 else 1
        table[x][y] = 1 - i
        table[y][x] = i
        nwon [(x, y)[i]] += 1
        nlost[(y, x)[i]] += 1
        count = search(pos + 1)
        nwon [(x, y)[i]] -= 1
        nlost[(y, x)[i]] -= 1
        table[x][y] = None
        table[y][x] = None
        if nmatch[y-1] == 0:
            return count
        else:
            # 1試合もしていないチームのうちxが勝つチームの選び方
            c = choose(n - y, (n-1)//2 - nwon[x])
            return c * count

    count = 0  # 組合せ数
    for i in range(2):
        table[x][y] = 1 - i
        table[y][x] = i
        nwon [(x, y)[i]] += 1
        nlost[(y, x)[i]] += 1
        if valid(x) and valid(y):
            count += search(pos + 1)
        nwon [(x, y)[i]] -= 1
        nlost[(y, x)[i]] -= 1
        table[x][y] = None
        table[y][x] = None
    return count

while True:
    n = int(input())
    if n == 0: break
    m = int(input())
    match = []
    for i in range(m):
        x, y = [int(t) for t in input().split()]
        match.append((x-1, y-1))
    print(solve(n, match))

File: 2018/d.c
#include <stdio.h>

#define MAX_N  9
#define MAX_MATCH  (MAX_N * (MAX_N - 1) / 2)

#define None  (-1)

struct Match {
  int x, y;
} match[MAX_MATCH];

int count;
int table[MAX_N][MAX_N];

int valid(int n, int x)
{
  int n0, n1;
  int y;
  n0 = n1 = 0;
  for (y = 0; y < n; y++) {
    if (table[x][y] == 0) { n0++; }
    if (table[x][y] == 1) { n1++; }
  }
  return n0 <= (n-1)/2 && n1 <= (n-1)/2;
}

void search(int n, int pos)
{
  int x, y, i;
  if (pos >= n * n) {
    count += 1;
    return;
  }
  x = pos / n;
  y = pos % n;
  if (x == y || table[x][y] != None) {
    search(n, pos + 1);
    return;
  }

  for (i = 0; i <= 1; i++) {
    table[x][y] = 1 - i;
    table[y][x] = i;
    if (valid(n, x) && valid(n, y)) {
      search(n, pos + 1);
    }
    table[x][y] = None;
    table[y][x] = None;
  }
}

int solve(int n, struct Match match[], int m)
{
  int i, j;
  count = 0;
  for (i = 0; i < n; i++) {
    for (j = 0; j < n; j++) {
      table[i][j] = None;
    }
  }
  for (i = 0; i < m; i++) {
    table[match[i].x-1][match[i].y-1] = 1;
    table[match[i].y-1][match[i].x-1] = 0;
  }
  for (i = 0; i < n; i++) {
    if (! valid(n, i)) { return 0; }
  }
  search(n, 0);
  return count;
}

int main()
{
  int i, n, m, x, y;
  for (;;) {
    scanf("%d", &n);
    if (n == 0) { break; }
    scanf("%d", &m);
    for (i = 0; i < m; i++) {
      scanf("%d%d", &x, &y);
      match[i].x = x;
      match[i].y = y;
    }
    printf("%d\n", solve(n, match, m));
  }
  return 0;
}
